File: modules/produtos.c
#include "../app/config.h"
#include <stdlib.h>
#include <stdio.h>

typedef struct {
    int id;
    char nome[INPUTSIZE];
    float valor;
} Produto;

static int proximo_produto = 0;

int cursor_proximo_produto(void) {
    return proximo_produto;
}

void imprimir_cabecalho_produto() {
    printf("  %3s %-49s %9s\n%s", "id", "nome", "valor",
    " ═══════════════════════════════════════════════════════════════════\n" );
}

void iprimir_produto(void *dado) {
    Produto * p= dado;
    printf("  %03d %-49s %9.2f\n", p->id, p->nome, p->valor );
}

int localizar_produto(void *a, void *b) {
    Produto *pa = a, *pb = b;
    return pa->id == pb->id ? FOUND : NOTFOUND;
}

int validar_produto(void *dado) {
    Produto * p= dado;

    //pode ser criada uma validacao personalizada
    return VALID;
}

Produto *alocar_produto(void) {
    Produto *produto = malloc(sizeof(Produto));
    if (!produto) {
        perror("malloc");
        return NULL;
    }

    // Valores padrao
    produto->id = proximo_produto += 1;
    produto->nome[0] = '\0';
    produto->valor = 0.0f;

    return produto;
}

File: modules/produtos.rs
use std::sync::atomic::{AtomicUsize, Ordering};
use std::fmt;

// ==========================
// Estrutura principal Produto
// ==========================
#[derive(Clone)]
pub struct Produto {
    pub id: usize,
    pub nome: String,
    pub valor: f32,
}

// ==========================
// Gerador de ID global
// ==========================
static PROXIMO_PRODUTO: AtomicUsize = AtomicUsize::new(0);

pub fn cursor_proximo_produto() -> usize {
    PROXIMO_PRODUTO.load(Ordering::SeqCst)
}

// ==========================
// Criação e validação
// ==========================
impl Produto {
    pub fn novo() -> Self {
        let id = PROXIMO_PRODUTO.fetch_add(1, Ordering::SeqCst) + 1;
        Produto {
            id,
            nome: String::new(),
            valor: 0.0,
        }
    }
}

pub fn validar_produto(_produto: &Produto) -> bool {
    // Pode ser personalizada depois
    true
}

// ==========================
// Impressão formatada
// ==========================
pub fn imprimir_cabecalho_produto() {
    println!(
        "  {:>3} {:<49} {:>9}\n{}",
        "id",
        "nome",
        "valor",
        " ═══════════════════════════════════════════════════════════════════"
    );
}

pub fn imprimir_produto(produto: &Produto) {
    println!(
        "  {:03} {:<49} {:>9.2}",
        produto.id,
        produto.nome,
        produto.valor
    );
}

// ==========================
// Busca e comparação
// ==========================
pub fn localizar_produto(a: &Produto, b: &Produto) -> bool {
    a.id == b.id
}
